### DISCUSSION TITLE
Asking feedback for prey predator estimation (with lots of plots!)

### DISCUSSION THREAD

**Discussion by hyunjimoon** [Data4DM/BayesSD#117] [brain belief 🟩]
May I ask for brief feedback on [this](https://github.com/Data4DM/stanify/blob/main/vignette/prey_predator_hier.ipynb) notebook? For instance, what explanation and plot do you think I can add that can be helpful for those not used to inference diagnostics (seminar students). Full list of possible plots [here](https://python.arviz.org/en/0.14.0/api/plots.html): 

@jandraor shared [avoidable error](https://royalsocietypublishing.org/doi/10.1098/rspb.2015.0347) paper with the following quote which would be a perfect ingredient to add to our #76.
- the validity of the statistical estimation procedure hinges on the independence of sequential measurement errors, which is clearly violated when observations are accumulated through time”
- models should be fit to raw, disaggregated data whenever possible and never to temporally accumulated data;
- when model assumptions, such as independence of errors, must be violated, careful checks for the effects of such violations should be performed;
- forecasts based on deterministic models, being by nature incapable of accurately communicating uncertainty, should be avoided
- stochastic models should be preferred to deterministic models in most circumstances because they afford improved accounting for real variability and increased opportunity for quantifying uncertainty. Post hoc comparison of simulated and actual data is a powerful and general procedure that can be used to distinguish model misspecification from real stochasticity.

**Comment by jandraor** [comment]
1. I simulate the model with some samples from the posterior and compare them against the data. This gives me a visual check of a "good fit".

2. [Pair Plot](https://python.arviz.org/en/0.14.0/api/generated/arviz.plot_pair.html#arviz.plot_pair) I find it very useful to evaluate the shape of the joint posterior distribution. Add correlations for more information. Lots of correlations indicate a difficult parameter space.

3. Trace plots.

4. For marginal distributions, I found it useful to overlap prior & posterior histograms. This plot tells how much information we gained from the inference process for a particular parameter.

**Reply by hyunjimoon** [reply]
Just to be clear, did you have a chance to look at https://github.com/Data4DM/stanify/blob/main/vignette/prey_predator_hier.ipynb mentioned above?..

**Reply by hyunjimoon** [reply]
Chobby obs time series from downsampling is apparent, but I don't know whether this is a feature or a bug.

**Reply by tomfid** [reply]
I'd call this a bug.
![image](https://user-images.githubusercontent.com/2151849/206276430-382817ea-9d95-4aa6-b623-ecee4bbd8220.png)
Not clear if it's a model issue or a solver issue though. Doesn't seem like it could be the sampler.

**Reply by hyunjimoon** [reply]
What makes you think it is a bug? Orange and blue are for the prey and red and green are for the predator.

**Reply by hyunjimoon** [reply]
Plot with a larger datapoint seems a little better.
<img width="1066" alt="image" src="https://user-images.githubusercontent.com/30194633/206300399-16ce25dd-bf84-47dd-a198-a1089e01337b.png">

**Reply by tomfid** [reply]
In the previous chart, it's oscillating with a period that appears to be 3*TIME STEP. I can't think of a legit way to generate something like that from the pred-prey model.

New chart seems much better.

**Reply by hyunjimoon** [reply]
TL;DR no problem in data, but problem in predictive plot code.

> oscillating with a period that appears to be 3*TIME STEP
Good observation.. that 3 (surprisingly) seem to be `R` (number of region). I ran R = 1 and 3 and time axis wrongly increased by three. Stanify stacked all spatial heterogeneity into temporal one! 
 
<img width="1346" alt="image" src="https://user-images.githubusercontent.com/30194633/206302555-58e774ac-da59-4e39-a484-fc3ac7873c4a.png">

As you pointed out last week, bug seems to be coming from stacking (chain, draw) as (prior_draw) in `draws2data()`. 
<img width="719" alt="image" src="https://user-images.githubusercontent.com/30194633/206306831-78837f8c-526b-4ff6-a712-8e88d8a247a1.png">

**Reply by tomfid** [reply]
Nice detective work! Good that it's not really a bug so much as a benign but confusing display choice.
